### DISCUSSION TITLE
Using `node-fetch` fails when proxy settings change

### DISCUSSION THREAD

**Discussion by matthewjamesadam** [microsoft/vscode-discussions#453] [Extension Development QnA] [answered]
Hello everyone,

We have a VSCode extension that uses `node-fetch`.  We also use proxy tools like Proxyman to debug http requests.  Proxyman works by providing a proxy at localhost:9090 that all traffic is routed through.  I've noticed that changing the proxy status while VSCode is open does not work:

* If I open Proxyman while VSCode is running, requests will not be shown in Proxyman (ie, requests are not being sent through the proxy).  Restarting VSCode makes this work correctly.
* If I close Proxyman while VSCode is running, all requests fail immediately with `reason: connect ECONNREFUSED 127.0.0.1:9090`.  ie, VSCode is not realizing that the proxy settings have changed, and is still trying to send requests through the proxy.

This is not so bad while using Proxyman for debugging, but we've noticed that some of our users are also running into this because some corporate VPN rules essentially work the same way -- they set default proxies for all traffic, and VSCode extensions don't update when the system proxy rules change.

I realize that VSCode is probably doing some work to set the proxy settings for extensions on startup, as by default node-fetch doesn't use the system proxy.  Is it expected that VSCode won't update when changing the system proxy while VSCode is running?  Is there any way we can get the proxy settings to update at runtime?

This is on macOS Monterey 12.5.1 with VSCode 1.75.1

Thanks everyone

**Comment by r3m0t** [comment]
Did you try `request-light` package?

**Reply by matthewjamesadam** [reply]
Thanks for the suggestion -- our app uses the `fetch` API heavily, is there a straightforward way to have `request-light` expose a fetch API?

**Reply by matthewjamesadam** [reply]
I've tried using `request-light` and it exhibited the same symptoms.  When the system proxy settings change while VSCode is running (ie, closing Proxyman), requests start to fail.\

One thing I did notice is that the request options for `request-light` take in an `agent` option of type `HttpAgent`.  I don't know what to supply for this, is it expected that VSCode extensions should need to supply a custom agent that monitors system changes?

**Comment by matthewjamesadam** [comment]
Note that this affects all plugins.  The below error is issued if I close Proxyman while VSCode is running.  I think this is a systemic VSCode issue, I have filed a bug in the VSCode repo for it here: https://github.com/microsoft/vscode/issues/174279

<img width="450" alt="Screen Shot 2023-02-13 at 11 30 19 AM" src="https://user-images.githubusercontent.com/2133518/218555525-f9d724c8-dfd2-43cc-93df-7f86cab2d01d.png">

**Comment by isidorn** [comment]
@chrmarti might be able to help here.

**Comment by chrmarti** [comment, marked as answer]
We cache the result of the proxy lookup for performance.
